rflx/model.py

```diff
@@ -1184,6 +1201,98 @@ def get_constraints(aggregate: Aggregate, field: Variable) -> Sequence[Expr]:
+                and isinstance(expression.left, First)
+                and expression.right == Number(1)
+            )
+
+        for name, expressions in self.checksums.items():  # pylint: disable=too-many-nested-blocks
+            if Field(name) not in self.fields:
+                self.error.append(
+                    f'checksum definition for unknown field "{name}"',
+                    Subsystem.MODEL,
+                    Severity.ERROR,
+                    name.location,
+                )
+
+            for e in expressions:
+                if not (
+                    isinstance(e, (Variable, Length))
+                    or (isinstance(e, ValueRange) and valid_lower(e.lower) and valid_upper(e.upper))
+                ):
+                    self.error.append(
+                        f'unsupported expression "{e}" in definition of checksum "{name}"',
+                        Subsystem.MODEL,
+                        Severity.ERROR,
+                        e.location,
+                    )
+
+                for v in e.findall(lambda x: isinstance(x, Variable)):
+                    assert isinstance(v, Variable)
+
+                    if Field(v.name) not in self.fields:
+                        self.error.append(
+                            f'unknown field "{v.name}" referenced'
+                            f' in definition of checksum "{name}"',
+                            Subsystem.MODEL,
+                            Severity.ERROR,
+                            v.location,
+                        )
+
+                if isinstance(e, ValueRange):
+                    lower = e.lower.findall(lambda x: isinstance(x, Variable))[0]
+                    upper = e.upper.findall(lambda x: isinstance(x, Variable))[0]
+
+                    assert isinstance(lower, Variable)
+                    assert isinstance(upper, Variable)
+
+                    if lower != upper:
+                        upper_field = (
+                            Field(upper.name) if upper.name.lower() != "message" else FINAL
+                        )
+                        lower_field = (
+                            Field(lower.name) if lower.name.lower() != "message" else INITIAL
+                        )
+                        for p in self._state.paths[upper_field]:
+                            if not any(lower_field == l.source for l in p):
+                                self.error.append(
+                                    f'invalid range "{e}" in definition of checksum "{name}"',
+                                    Subsystem.MODEL,
+                                    Severity.ERROR,
+                                    e.location,
+                                )
+
+        checked = [
+            e.prefix.identifier
+            for e in self.field_condition(FINAL).findall(lambda x: isinstance(x, ValidChecksum))
+            if isinstance(e, ValidChecksum) and isinstance(e.prefix, Variable)
+        ]
```

Comment:
Wouldn't `checked = set(e.prefix.identifier for ...)` be faster than creating a list and converting it to a set?

Reply:
Probably, changed in 8f7a30b.